### PR TITLE
Feature/enabling by overload

### PR DESCRIPTION
**Pull request by tobias-loew**

### CODE REVIEW

File: include/boost/flags.hpp [hunk 1]
```diff
@@ -42,15 +42,15 @@
         olives      = boost::flags::nth_bit(3), // == 0x08
     };
     // enable Boost.Flags for pizza_toppings
-    template<> struct boost_flags_enable<pizza_toppings> : std::true_type {};
+    static constexpr bool boost_flags_enable(pizza_toppings{}) { return true; }
```
Comment (andreasbuhr) [inline include/boost/flags.hpp]: I think the curly braces at the argument are wrong. "pizza_toppings" is the type here. Just remove the curly braces.
... but this is documentation, which is unfinished, as you said.

File: include/boost/flags.hpp [hunk 2]
```diff
@@ -42,15 +42,15 @@
...
 
     enum class ice_cream_flavours : unsigned int {
         vanilla     = boost::flags::nth_bit(0), // == 0x01
         chocolate   = boost::flags::nth_bit(1), // == 0x02
         strawberry  = boost::flags::nth_bit(2), // == 0x04
     };
     // enable Boost.Flags for ice_cream_flavours
-    template<> struct boost_flags_enable<ice_cream_flavours> : std::true_type {};
+    static constexpr bool boost_flags_enable(ice_cream_flavours{}) { return true; }
```
Comment (andreasbuhr) [inline include/boost/flags.hpp]: Remove curly braces

**Comment by andreasbuhr** [comment]
godbolt-example with this branch: https://godbolt.org/z/zMP5zefh3

**Comment by andreasbuhr** [comment]
There is actually a drawback to this approach. Consider you have a typo and you have
```
template<> struct boost_enable_flags<ice_cream_flavours> : std::true_type {};
```
In the old approach, this led to an error message.

For the new approach, this would be 
```
static constexpr bool boost_enable_flags(ice_cream_flavours) { return true; }
```
which does not lead to any error message.

A similar problem arises with the `#include <boost/flags.hpp`. If you have a large project, you don't want to include flags.hpp everywhere, so you include it from some central header file. You then assume it is available everywhere. The you define
```
static constexpr bool boost_flags_enable(ice_cream_flavours) { return true; }
```
and everything seems to be fine, when in fact, nothing happened because you lack the include.

So it would be nice if you could provide some way to make sure the registration worked. Something like
`static_assert(boost::flags::is_registred(ice_cream_flavours));`.
However, this makes the usage of this library a bit ugly. You then have two lines of code for each enum-registration.

File: include/boost/flags.hpp (hunk 1 above)
Comment (tobias-loew) [reply include/boost/flags.hpp]: you're right - I removed the curly braces

**Comment by tobias-loew** [comment]
> So it would be nice if you could provide some way to make sure the registration worked. Something like `static_assert(boost::flags::is_registred(ice_cream_flavours));`. However, this makes the usage of this library a bit ugly. You then have two lines of code for each enum-registration.

You can already write
`static_assert(boost::flags::enable<ice_cream_flavours>::value, "ice_cream_flavours not enableb for Boost.Flags");`
which would static-assert in the first case, and cause a compilation error, when the include is missing.

Yes, it is a small drawback that typos with `boost_flags_enable` do not cause errors, but
1. You'll get an error on the first flag-like usage of your enum - hopefully in your tests ;)
2. You can still use the "old" enabling style: `template<> struct boost::flags::enable<my_enum> : std::true_type{};`
3. `boost_flags_enable` has the hugh advantages 
    - it can be defined in the same namespace as the enum (e.g. right after it) - no need to close and reopen the current namespace, which is very annoying
    - you can enable it even within classes by using a `friend` function (no way to do this with template specialization)

IMHO, if you want to be on the safe-side, you could write
```
static constexpr bool boost_flags_enable(my_enum) { return true; }
static_assert(boost::flags::enable<my_enum>::value, "");
```

**Comment by andreasbuhr** [comment]
Maybe you can add the option to do 
```
static constexpr bool boost_flags_enable(my_enum) { return true; }
static_assert(boost::flags::enable<my_enum>::value, "");
```
to the documentation. I think it's essential for the following user story:

Assume you have a few million lines of code which you'd like to improve using boost::flags. The approach would be to:
1. Include boost/flags.hpp in a central place
2. Identify enums which seem to be flags, say you find around 30 different enums of this kind.
3. Add the code snippet above for each enum
4. Fix all compilation errors which arise.

In this story, there is no "first flag-like usage" of your enum, because your codebase compiled just fine before. So any spelling mistake in boost_flags_enable without the static_assert() would just go unnoticed.

**Pull request merged**